Implementation of Category

oleksii-yavtushenko opened the pull request.

oleksii-yavtushenko committed:
Change List<Category> connection to @ManyToMany
Change CascadeType to DETACH
Add @EqualsAndHashCode (callSuper = true)

oleksii-yavtushenko committed:
Add @EqualsAndHashCode (callSuper = true)

oleksii-yavtushenko committed:
Override hashcode() and equals(Object).

oleksii-yavtushenko committed:
Created CategoryDTO.

oleksii-yavtushenko committed:
Refactor ProductDTO.

oleksii-yavtushenko committed:
Add id-field to ordering_inventory and product_category tables.

oleksii-yavtushenko committed:
Add properties to show sql-queries of hibernate.

oleksii-yavtushenko committed:
Create abstract EntityMapper to map Entities with their DTO objects.

oleksii-yavtushenko committed:
Implement EntityMapper with ProductMapper.

oleksii-yavtushenko committed:
Implement EntityMapper with CategoryMapper.

oleksii-yavtushenko committed:
Created MappingException to throw exceptions connected with entity mapping.

oleksii-yavtushenko committed:
Implement ProductService.

oleksii-yavtushenko committed:
Change Category.parent from @OneToOne to @ManyToOne relation.

oleksii-yavtushenko committed:
Created AbstractDTO - superclass of all DTO objects.

oleksii-yavtushenko committed:
Extend with AbstractDTO.

oleksii-yavtushenko committed:
Change strategy of @GeneratedValue to IDENTITY.

oleksii-yavtushenko committed:
Refactor EntityMapper and its implementations:
Add mapEntity(ProductDTO, Product);
Add mapDTO(Product, ProductDTO);
Make mapEntity(ProductDTO), mapDTO(Product) capsules of mapEntity(ProductDTO, Product), mapDTO(Product, ProductDTO) accordingly.

oleksii-yavtushenko committed:
Create Validator to validate Entity and DTO objects.

oleksii-yavtushenko committed:
Refactor ProductService:
- Wire Validator to ProductService;
- Add validation with Validator;
- Make ProductDTO object update from database in ProductService#save() method.

oleksii-yavtushenko committed:
Add limitation to DTO Generics.

oleksii-yavtushenko committed:
Refactor getById() to interact with ProductMapper.

oleksii-yavtushenko committed:
Edit equals(Object) to get equality not only with id.

oleksii-yavtushenko committed:
Add parametrized constructor to accept also id.

oleksii-yavtushenko committed:
Change primitives to their capsule-analog in parametrized constructor.

oleksii-yavtushenko committed:
Minor refactoring.

oleksii-yavtushenko committed:
Refactor equals.

oleksii-yavtushenko committed:
Add own overridden equals(Object) and hashcode().

oleksii-yavtushenko committed:
Fix bug.

oleksii-yavtushenko committed:
Add feature to return updated product in update(ProductDTO).

oleksii-yavtushenko committed:
Refactor getById(Integer) to return null if product is not present in the database.

oleksii-yavtushenko committed:
Delete of overridden equals(Object) and hashcode().

oleksii-yavtushenko committed:
Create Validator - interface of all validators.

oleksii-yavtushenko committed:
Add ValidationException that throws in Validator implementations.

oleksii-yavtushenko committed:
Add ProductValidator to validate ProductDTO objects.

oleksii-yavtushenko committed:
Moved old Validator realization to GeneralValidator with minor refactoring.

oleksii-yavtushenko committed:
Delete of unused import.

oleksii-yavtushenko committed:
Refactoring connected with change of validation architecture.

oleksii-yavtushenko committed:
Create and implement ProductServiceTest.
100% code coverage of testing ProductService provided.

Minor refactoring.

oleksii-yavtushenko committed:
Implement Category Controller, Service, Mapper, Validator and Tests.

Implement Category + Code Coverage of controller and service.

Implement CategoryService.

Create and implement CategoryValidator.

Add Component annotation.

Edit URL-path of deleteCategoryById.

Implement mapUpdate.

Cover CategoryController with 100% code coverage tests.

Change type of exception of not implemented methods.

Create CategoryService.

Create and implement CategoryController.

oleksii-yavtushenko committed:
Refactor of the ProductService Implementation.

oleksii-yavtushenko committed:
Refactor some classes.

oleksii-yavtushenko committed:
Minor changes.

oleksii-yavtushenko committed:
Merge branch 'master' into category-impl

MaxAbrosimov commented on src/test/java/services/ProductServiceTest.java:
```diff
@@ -159,18 +165,20 @@ public void update_ObjectIsPresentInTheDatabaseAndDataIsValid_ShouldReturnUpdate
 
-    @Test(expected = NoSuchElementException.class)
+    @Test(expected = ServiceException.class)
     public void update_ObjectWithIDIsNotPresent_ShouldThrowNoSuchElementException() throws ServiceException {
```
Please take a look on ExpectedException approach

MaxAbrosimov commented on src/main/java/com/griddynamics/service/CategoryService.java:
```diff
@@ -0,0 +1,100 @@
+        Optional<Category> optCategory = categoryRepository.findById(id);
+
+        if (!optCategory.isPresent()) {
+            throw new ServiceException("Product with " + id + " is not present in the database.");
```
Please use String.format instead of concatenation

oleksii-yavtushenko committed:
Little changes related to Changes Request

Refactor tests expectation of the exceptions.

Change concatenation to String.format() in some exception messages.

Add annotation Repository to the repositories.

MaxAbrosimov commented on src/main/java/com/griddynamics/service/CategoryService.java:
```diff
@@ -0,0 +1,100 @@
+
+        CategoryDTO categoryDTO = null;
+
+        if (optCategory.isPresent()) {
```
please take a look on orElseGet method in optionals

MaxAbrosimov commented on src/main/java/com/griddynamics/service/CategoryService.java:
```diff
@@ -0,0 +1,100 @@
+
+        Optional<Category> optCategory = categoryRepository.findById(categoryDTO.getId());
+
+        if (!optCategory.isPresent()) {
```
same for [orElseThrow](https://www.baeldung.com/java-optional-throw-exception#optionalorelsethrow)

MaxAbrosimov commented on src/test/java/services/CategoryServiceTest.java:
```diff
@@ -0,0 +1,230 @@
+    private CategoryService categoryService;
+
+    @Before
+    public void beforeAllMethods() {
```
look like begore every method

MaxAbrosimov replied:
rename to init/setup or something like that

MaxAbrosimov commented on src/test/java/services/CategoryServiceTest.java:
```diff
@@ -0,0 +1,230 @@
+        /* Then */
+
+        try {
+            categoryService.update(updateCategory);
```
Please take a look on ExpectedException class

oleksii-yavtushenko committed:
Minor refactoring.
